### DISCUSSION TITLE
Average Visit Time Calculation

### DISCUSSION THREAD

**Discussion by talthanas** [umami-software/umami#1816] [Ideas]
I'm testing on a site where users load a web page with inline video and often stay for long periods of time.
Running Umami and Google Analytics in parallel has Visitor and Page counts in the same general ballpark but Visit Duration in Umami is 1-2 minutes where Google Analytics shows as 40min to over 1 Hours averages. The longer duration is more correct for this site.

Is there a default timeout or delta between page views that can be tweaked in code or possible exposed in configuration?

To experiment, I added client side code to register an event every minute to see if that had any effect on average visit duration and it did not appear to even though Umami did track thousands of events.

**Comment by mikecao** [comment]
The visit time calculation is based on the delta between page visits. So if a user comes to a page, watches a video, and doesn't visit any other pages, we can't tell where there visit ended. We would probably have to come up with a solution to "ping" umami with the updated time while the user sits on the page.

**Comment by talthanas** [comment]
Totally makes sense.
Thoughts/questions: 

Can an event be taken into account or act as that "ping" vs just the  page view? 

Or could there be a configuration setting that exposes that max delta time? I'm taking a wild guess that it's around 5 minutes just looking at some of my reports.

**Comment by gnarlex** [comment]
I'm not completely familiar with the analytics script. But couldn't we add a `visibilitychange` handler in the analytics script and use [`Navigator.sendBeacon()`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon) there to send information to umami once the user leaves a page?

MDN provides following example:
```js
document.addEventListener('visibilitychange', function logData() {
  if (document.visibilityState === 'hidden') {
    navigator.sendBeacon('/log', analyticsData);
  }
});
```

**Comment by Sammy-T** [comment]
To add to @gnarlex's suggestion, initial page load and `visibilityState`'s 'visible' and 'hidden' values can be used to track when the page enters and leaves focus. I think it could be feasible to send this delta to the server-side. This should handle the elapsed time of each view when only visiting one page, navigating between site pages, or navigating between browser tabs.

**Comment by Sammy-T** [comment]
One possible concern is that the tracker's current `XMLHttpRequest` appears to cache the server's response whereas `Navigator.sendBeacon()` would only be able to respond **true** or **false** to show if the request was queued.

**Comment by ryancanzo** [comment]
Are there any plans to update the method for calculating the visit time in the near future? We're in a similar situation as we have a web app that people can use for long periods without navigating between different pages. We use events for various actions on the page so taking these into account for the visit time would be useful. Thanks

**Comment by rohandebsarkar** [comment]
I have been playing around with this a bit. I think the solution is,

1. Create a new column in the database, in the session table, called `ended_at`.
2. Then a new collection type, called `sessionend`.
3. In the tracker, add an event listener for visibility change and send a `sessionend` request if the visibility state is hidden. (mentioned previously by @gnarlex, but using fetch keepalive).
4. When `sessionend` request is received, change the session table to update the `ended_at` for the particular session.

Then the visit time can be calculated using the delta between `created_at` and `ended_at` for the particular session.

@mikecao, could this be a possible solution?

**Comment by imjuniper** [comment]
Doesn't the visibility change event also fire when a tab gets inactive? I wonder if it would duplicate views or something in cases where the user leaves the tab then comes back.

**Comment by rohandebsarkar** [comment]
> Doesn't the visibility change event also fire when a tab gets inactive? I wonder if it would duplicate views or something in cases where the user leaves the tab then comes back.

@jakobbouchard, I don't think that will be a problem, since every time the visibility change occurs (hidden), a request will be fired. And each time the timestamp will be updated in the database. The last timestamp value will only be in the record, the other values will be discarded.

**Comment by imjuniper** [comment]
@rohandebsarkar Oh yeah that makes sense!

**Comment by timbomckay** [comment]
Has there been any update on this? Seems like currently a user has to visit another page in order to calculate visit time.

**Comment by VogelDevelopment** [comment]
For me it's also an annoying problem - my user-time-calculations are totally off because of this. So +1 for fixing this issue and f.e. going back to sendBeacon - it's critical to know how much time users spent on each page, specially before leaving the website. Makes the difference between "this page is too boring, I lost interest" and "okay that gave me all information I need"

**Comment by Kovah** [comment]
I also created #3287 which would solve this.
@mikecao Would you be interested in a pull request for this? Because at least for me it's a deal-breaker and it looks like others would like to have this too.

**Reply by mikecao** [reply]
We are brainstorming a few ways to do this. The issue is that Clickhouse has a different data structure than MySQL/PostgreSQL and we need to find a solution that works on both. Was thinking about a /ping endpoint that would capture activity while the user is still on the page.

**Reply by Kovah** [reply]
Any news on this?

**Reply by maxiride** [reply]
The idea of a ping, or to better say it a heartbeat, would fit very nicely with https://github.com/umami-software/umami/discussions/2544#discussioncomment-8636483, https://github.com/umami-software/umami/discussions/3287 and https://github.com/umami-software/umami/issues/3518

I mean, there is some traction from the community on the subject.

**Reply by kreativekrise** [reply]
There is a similar discussion https://github.com/rybbit-io/rybbit/discussions/348. Perhaps their ideas and findings will help develop a reliable solution. I tested some other products, like Plausible and Rybbit, and none of them track time for single-page visits. This would give Umami a chance to stand out.

**Comment by kreativekrise** [comment]
I would very much like to switch to Umami. Unfortunately, without this metric, it is not possible. The issue is already four years old. Will there be a solution in the near future?
